LanguageFeatures/Horizontal-inference/bounds_t01.dart

```diff
@@ -0,0 +1,29 @@
+// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion Horizontal inference allows certain arguments of an invocation
+/// (specifically function literals) to be type inferred in a context derived
+/// from the type of other arguments.
+///
+/// @description Checks horizontal inference when type bounds specified
+/// @author sgrekhov22@gmail.com
+
+import "../../Utils/expect.dart";
+
+class C1 {}
+class C2 extends C1 {}
+class C3 extends C2 {}
+
+void f1<T extends C1>(void Function(T) a, T b) {
+  Expect.equals(T, C3); // Without horizontal inference it'd be C1
+}
+
+void f2<T extends C2>(T Function() a, void Function(T) b) {
+  Expect.equals(T, C3); // Without horizontal inference it'd be C2
+}
+
+main() {
+  f1((v){}, C3());
+  f2(() => C3(), (x) {});
```

Comment:
I actually can't make it anything other than C3 in any version down to 2.9.0-21.0.dev (tried about 5 versions).

Looking a bit more at Paul's example:

```dart
void main() {
  Iterable<int> values = [1, 2];
  var largestValue = values.fold(0, (a, b) => a < b ? b : a); // Compile-time error.
  // ...
}
```

It is crucial that the invocation of `fold` does not have a context type. If we declare `int largestValue` then the context type of `int` will guide the inference to choose `int` as the actual type argument to the invocation of `fold`, and then everything just works.

So one thing to be aware of in these tests is that they should not provide any context types when we expect to see an improvement, or at least: We must be very careful if we provide a context type, because it might very well eliminate the need for the horizontal inference altogether.

I can see that the occurrence of `C3()` causes the inference to infer the actual type argument `C3` both with and without the `inference-update-1` feature, so we do have a situation here where `inference-update-1` is not being tested.

The effect of `inference-update-1` is actually not to change the inferred type argument of `f1` and `f2`, at least not in an example like this one. The change amounts to using the inferred type argument `C3` in the context type provided for the function literal that gets to be inferred later (the one in stage 1, not 0). This means that `inference-update-1` makes a difference if we're using the type of the parameter of the function literal.

So let's change `C3` such that we can see whether a given object is a `C3` or something less than that:

`class C3 extends C2 { void m() {}}`

With that, we can use `f1((v) => v.m(), C3())` and `f2(() => C3(), (x) => x.m())`.

I believe that would change the test such that it does test `inference-update-1`, rather than just testing a behavior which is already present in older versions of the language.

Reply:
Erik thank you. I fixed the tests. Please re-review. Note that a couple of the tests produce the same result with and without horizontal inference (with `no-inference-update-1` and without it). These are "negative" tests, that checks cases when there are no inference. For example `explicit_type_arguments_t01.dart` checks that if type arguments specified then no horizontal inference happened.

Reply:
All great, thanks!